### ISSUE TITLE
CustomScrollView + SliverAppBar + WebView

### ISSUE THREAD

**Issue by timrijckaert** [flutter/flutter#31243]
Hi, 

When trying to implement a floating `SliverAppBar` in conjunction with a `WebView` Flutter drops a significant amount of frames.
It seems like the WebView is resizing itself whenever the `SliverAppBar` is scrolled out of view.

```
@override
Widget build(BuildContext context) {
  return CustomScrollView(
    slivers: <Widget>[
      SliverAppBar(
        title: const Text("Heading"),
        floating: true,
      ),
      SliverFillRemaining(
        child: WebView(initialUrl: "http://stackoverflow.com"),
      )
    ],
  );
}
```

As seen on this GIF:

![flutter_1](https://user-images.githubusercontent.com/6739443/56343029-ecba7f00-61b9-11e9-963a-4e2bc3836f77.gif)

**Comment by Sun3** [comment]
I am also have the same issues but for me the full page article is not shown and gets cut off especially if you reload the page.

**Comment by dcov** [comment]
Yup that's exactly what's happening. `SliverFillRemaining` uses the `remainingPaintExtent` constraint, which is the extent that's yet to be painted on, to size its child. This changes as you scroll, which causes the `AndroidView`/`WebView` to resize as you noticed.

`SliverFillViewport` is what you're looking for. Although it still doesn't solve your problem of nested scrolling, which I don't think is possible as of now.

**Comment by Sun3** [comment]
I just tested with the ```SliverFillViewport``` and like you the end result is the same. No difference with the ```SliverFillRemaining```. Short pages show up but longer pages they get cut off.

**Comment by Reekoooo** [comment]
partial solution in my repo

I will share my view here repeating what I wrote in my answer in [stackoverflow](https://stackoverflow.com/a/55738621/10660823) may be it can help providing a solution

I can give you my view for why it is not possible to do that with the current state of the plugin ,by default the Webview only respond to drag gesture when no other views claim that gesture. On the other hand, Scrolling slivers like SliverList, which is needed to make the SliverAppBar to scroll up ,by default consumes all drag scrolling gestures -although you can disable that by providing noScrollPhysics - but once the WebView cover all the screen there is actually noway to report back to the slivers to start consume scroll again .

So the solution is to modify the WebView plugin itself to provide a callback

I already did that just for a proof of concept you can check my repo [Plugins](https://github.com/Reekoooo/plugins/tree/master/packages) by adding a new callback to the webView 
it will report back the ScrollY position which you can use to control your nested scroll

**Comment by lizhuoyuan** [comment]
the same issue

**Comment by do4Mother** [comment]
Same issues in here. I try 3 plugin webview. But can't use sliverappbar. My app cant scroll.

**Comment by phoenixsky** [comment]
same issue + 1

**Comment by conghaonet** [comment]
same issue：NestedScrollView + SliverAppBar + WebView

**Comment by stevemoreau** [comment]
To anyone following this issue, I'm getting into this, and will try to add some more information in a near future (hopefully, if I could understand it).

Meantime, the pending POC commit of @Reekoooo (see above) : https://github.com/Reekoooo/plugins/commit/666b35ff50fa45991733999e1e6dcfdbf5a19be5

**Gesture arena of a valid use case**
Standard AppBar (fixed) + Tabview with 5 tabs (so horizontal scrolling possible)
_OUTPUT BEHAVIOR_: The WebView allows scroll from the start to the end of the HTML

```
10:26:38.532 1 info flutter.tools I/flutter (19503): Gesture arena 1    ❙ ★ Opening new gesture arena.
10:26:38.557 2 info flutter.tools I/flutter (19503): Gesture arena 1    ❙ Adding: Instance of '_CombiningGestureArenaMember'
10:26:38.567 3 info flutter.tools I/flutter (19503): Gesture arena 1    ❙ Accepting: Instance of '_CombiningGestureArenaMember'
10:26:38.577 4 info flutter.tools I/flutter (19503): Gesture arena 1    ❙ Adding: LongPressGestureRecognizer#cb2b8(debugOwner: GestureDetector, state: ready)
10:26:38.587 5 info flutter.tools I/flutter (19503): Gesture arena 1    ❙ Adding: HorizontalDragGestureRecognizer#26c36(start behavior: start)
10:26:38.597 6 info flutter.tools I/flutter (19503):                    ❙ HorizontalDragGestureRecognizer#26c36(start behavior: start) calling onDown callback.
10:26:38.607 7 info flutter.tools I/flutter (19503): Gesture arena 1    ❙ Closing with 3 members.
10:26:38.611 8 info flutter.tools I/flutter (19503): Gesture arena 1    ❙ Eager winner: Instance of '_CombiningGestureArenaMember'
10:26:38.620 9 info flutter.tools I/flutter (19503):                    ❙ HorizontalDragGestureRecognizer#26c36(start behavior: start) calling onCancel callback.
```

**Gesture arena of a invalid use case (partial HTML, then cut or infinite height)**
Scaffold with SliverAppBar + Webview without EagerGestureRecognizer in `gestureRecognizers` property
_OUTPUT BEHAVIOR_: SliverAppBar correctly shrinked on scroll global scroll, but partial webview scroll (HTML wrapped to the height of the screen, and the WebView does not allow scroll from the start to the end of the HTML)

```
09:54:20.909 1 info flutter.tools I/flutter (19503): Gesture arena 5    ❙ ★ Opening new gesture arena.
09:54:20.913 2 info flutter.tools I/flutter (19503): Gesture arena 5    ❙ Adding: Instance of '_CombiningGestureArenaMember'
09:54:20.913 3 info flutter.tools I/flutter (19503): Gesture arena 5    ❙ Adding: LongPressGestureRecognizer#adde2(debugOwner: GestureDetector, state: ready)
09:54:20.913 4 info flutter.tools I/flutter (19503): Gesture arena 5    ❙ Adding: VerticalDragGestureRecognizer#892bd(start behavior: start)
09:54:20.913 5 info flutter.tools I/flutter (19503):                    ❙ VerticalDragGestureRecognizer#892bd(start behavior: start) calling onDown callback.
09:54:20.913 6 info flutter.tools I/flutter (19503): Gesture arena 5    ❙ Closing with 3 members.
09:54:20.938 7 info flutter.tools I/flutter (19503): Gesture arena 5    ❙ Rejecting: LongPressGestureRecognizer#adde2(debugOwner: GestureDetector, state: possible)
09:54:20.941 8 info flutter.tools I/flutter (19503): Gesture arena 5    ❙ Accepting: VerticalDragGestureRecognizer#892bd(start behavior: start)
09:54:20.941 9 info flutter.tools I/flutter (19503): Gesture arena 5    ❙ Self-declared winner: VerticalDragGestureRecognizer#892bd(start behavior: start)
09:54:20.941 10 info flutter.tools I/flutter (19503):                    ❙ VerticalDragGestureRecognizer#892bd(start behavior: start) calling onStart callback.
09:54:20.948 11 info flutter.tools I/flutter (19503):                    ❙ VerticalDragGestureRecognizer#892bd(start behavior: start) calling onUpdate callback.
09:54:21.059 12 info flutter.tools I/flutter (19503):                    ❙ VerticalDragGestureRecognizer#892bd(start behavior: start) calling onUpdate callback.
09:54:21.123 13 info flutter.tools I/flutter (19503):                    ❙ VerticalDragGestureRecognizer#892bd(start behavior: start) calling onUpdate callback.
09:54:21.151 14 info flutter.tools I/flutter (19503):                    ❙ VerticalDragGestureRecognizer#892bd(start behavior: start) calling onUpdate callback.
```

Next things to investigate for me `VerticalDragGestureRecognizer` and arena election process, and `SliverList` internals to check how it can properly scroll its own view and let the scroll to be handle by the `VerticalDragGestureRecognizer` which (I currently think) is the top gesture handler for Sliver based layouts.

Please advise if it is not the right approach.

**Comment by stevemoreau** [comment]
> same issue：NestedScrollView + SliverAppBar + WebView

Passing by, add NestedScrollView + SliverAppBar + WebView previously, and recently moved to CustomScrollView + SliverAppBar + WebView because it would crash in **release** builds only.

**Comment by liyuqian** [comment]
@iskakaushik, @amirh  : is this also a known performance issue that will be solved once the thread-merging work is done for the WebView?

**Comment by timrijckaert** [comment]
@liyuqian Is this a PR you are referencing that we can follow up upon?

**Comment by mklim** [comment]
This affects all platform views, not just webview. I expect that `google_maps_flutter` would also have this problem. @liyuqian I don't think this has anything to do with threading logic. We do resize the a platform view whenever we detect that the screen size has changed. The resize operation is extremely heavy because it involves detaching and creating a new virtual display of the new size to attach the platform view to. The expected behavior is to resize to make room for things like the system soft keyboard.

At a first glance I was going to say that it seems like the way we detect whether or not we should resize is broken, but I think there may be some complexity here with `SliverAppBar` actually effectively changing WebView's size as it moves that makes this especially difficult.

**Comment by ollisal** [comment]
This other webview plugin https://pub.dev/documentation/flutter_inappbrowser/latest/ has a scroll callback that works on both iOS and Android, similarly to @Reekoo Android PoC for webview_flutter.

Not sure if it works inside a NestedScrollView though to achieve the behavior that's attempted here.

Currently I have a setup where I just change the height of a top bar manually from the scroll callbacks of the InAppWebView. To solve the performance issue described here from resizing the webview, I made the webview constant size instead, with the top bar covering part of it in a Stack. This was possible because I'm showing a partially customized web page where I can add a top padding matching the covered-up part.

**Comment by mrifni** [comment]
I am having this issue, whats the update on this ? why the issue is closed?

**Comment by mrifni** [comment]
Any updates on this ? I am having the same issue.

**Comment by sdlu1992** [comment]
I am having the same issue, Please fix the bug that webview will resizing itself and out  the boundary of itself

**Comment by mathieusachet** [comment]
same problem here

**Comment by khawlaMS** [comment]
same problem with me ... please give us any solution

**Comment by Sly2024** [comment]
same problem here

**Comment by amirh** [comment]
Is this seen on Android only or on iOS as well?
As Michael mentioned, resizing a platform view on Android is pretty expensive, though I'm not sure that without further investigations I'd blame the jank seen in the attached gif on the VD resize as it seems like the Flutter frame itself is janky.

Anyway this needs further investigation, and hopefully we can make this a priority soon enough.  Unfortunately I cannot guarantee a timeline as this doesn't currently make the priority cut for the Ecosystem/Plugins team.

**Comment by liyuqian** [comment]
Not sure if this is Android only or problematic on iOS as well.

@blasten : since you've recently added platform_views_scroll_perf__timeline_summary and platform_views_scroll_perf_ios__timeline_summary perf test (https://github.com/flutter/flutter/pull/52839), I wonder if we can have a perf test for this case too. Then we'll be sure whether iOS also has problems, and track our fix on those platforms.

**Comment by blasten** [comment]
@liyuqian that sounds good! This is most likely Android specific since as @amirh mentioned resizing platform views is very expensive on Android.

I haven't profiled this yet, but I presume hybrid composition on Android will fix this issue.

**Comment by heshaShawky** [comment]
Same issue here too!!

**Comment by mrifni** [comment]
this is a year old issue, would be good to know any roadmap to fix this issue instead hanging developers and not giving an update.

**Comment by bilaloz** [comment]
Any updates on this ? I am having the same issue.

**Comment by BeomJobs** [comment]
Any updates on this ? I am having the same issue.

**Comment by timrijckaert** [comment]
People if you have the same issue don't forget to thumbs up this issue.
Flutter team triage issues by the amount of thumbs up an issue has.

Thanks

**Comment by darshankawar** [comment]
Issue reproducible on latest dev channel and with latest `webview_flutter`.

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel dev, 1.20.0-0.0.pre, on Mac OS X 10.15.2 19C57, locale en-US)
    • Flutter version 1.20.0-0.0.pre at /Users/deeptibelsare/Documents/Fluttersdk/flutter
    • Framework revision d9653445f4 (8 days ago), 2020-06-09 18:43:03 -0400
    • Engine revision e8c13aa012
    • Dart version 2.9.0 (build 2.9.0-14.0.dev 5c1376615e)

 
[!] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/deeptibelsare/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/deeptibelsare/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/deeptibelsare/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    ✗ Android license status unknown.
      Try re-installing or updating your Android SDK Manager.
      See https://developer.android.com/studio/#downloads or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions.

 
[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.9.3

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.1.0 (API 27) (emulator)

! Doctor found issues in 1 category.
```

</details>

**Comment by keluokeda** [comment]
same problem here

**Comment by AmbrocioIsaias2808** [comment]
well i have the same issue

**Comment by qingweiSun** [comment]
well i have the same issue any new info on this?

**Comment by akashicrecorder** [comment]
same problem here

**Comment by DFelten** [comment]
Same issue here. When using a WebView it's laggy and not using the height of the WebView. Instead only the height of the screen.

**Comment by virenvvarasadiya** [comment]
A little bit different case but i face same issue with this package. I am using webview in pageview, when I use               allowImplicitScrolling: true property it is lagging too much.

**Comment by liyuqian** [comment]
@blasten : is there any progress on this?

**Comment by blasten** [comment]
@liyuqian no progress on this yet. Resizing webviews while using virtual displays is very expensive.  The other approach I've been working on makes the resize less expensive, but causes other perf regressions see: https://github.com/flutter/flutter/issues/62303

**Comment by blasten** [comment]
@timrijckaert have you thought about having a fixed size container while the animation is in progress and eventually resizing the container, but only at the end of the animation?

**Comment by blasten** [comment]
@timrijckaert have you thought about having a fixed size container while the animation is in progress and eventually resizing the container, but only at the end of the animation?

@Piinks do you know if it possible to achieve something like that with `CustomScrollView`?

**Comment by timrijckaert** [comment]
We simply did not investigate this any further because of the performance issues.
Feel free to provide an example to help out the others.

**Comment by blasten** [comment]
The performance issue seems to be fixed by hybrid composition, although there's a separate issue with the gesture disambiguation that I'm trying to figure out.

**Comment by blasten** [comment]
@yjbanov this is the type of issues where having declarative gesture disambiguation would simplify or eliminate most of our current hacks.

**Comment by blasten** [comment]
@bparrishMines what about changing the webview example app to include "CustomScrollView + SliverAppBar + WebView"?  

We don't have a good way to test this e2e, so by having it as the example app, we are forced to test it manually whenever someone opens the app.

**Comment by blasten** [comment]
This has been fixed in Flutter 1.22/ WebView 1.0 - use hybrid composition in the Flutter WebView plugin.

https://github.com/flutter/plugins/tree/master/packages/webview_flutter#using-hybrid-composition

**Comment by hanniiel** [comment]
I'm trying to implement CustomScrollView with SliverAppBar. I'm using SurfaceAndroidWebView the header scrolls with the webview that works just fine but when the header is collapsed the scroll doesn't scroll over the webview content and as a result the full webview content isn't visible.

before this update I achieved this functionality by injecting some js into the webview content to detect the scroll direction into the html content and by listening if the body top was visible then I changed the gestureRecognizers according to those events(exmaple:https://www.facebook.com/1792131602/videos/10213309766977319/ ) but It seems this update doesn't fix that or am I missing something?

best regards

`@override
  void initState() {
    super.initState();
    if (Platform.isAndroid) WebView.platform = SurfaceAndroidWebView();

    _scrollController = ScrollController();

  }
return CustomScrollView(
      slivers: <Widget>[
        SliverAppBar(
          primary: false,
          title: const Text("Heading"),
          floating: true,
          toolbarHeight: 200,
          flexibleSpace: FlexibleSpaceBar(),
        ),
        SliverFillRemaining(
          child: WebView(
            initialUrl: "https://soompi.com",
            javascriptMode: JavascriptMode.unrestricted,
          ),
        )
      ],
    )`

using:
webview_flutter 1.0.3

Flutter 1.22.2 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 84f3d28555 (3 days ago) • 2020-10-15 16:26:19 -0700
Engine • revision b8752bbfff
Tools • Dart 2.10.2

**Comment by heshaShawky** [comment]
Still the issue not fixed!! using **Flutter 1.22.3** and **flutter_webview 1.0.7** with enabling the **Hybrid Composition** on every **layout possible**: 

1. Using `CustomScrollView` with 'SliverAppBar' and `SliverFillRemaining` 
2. or using `NestedScrollView` with 'headerSliverBuilder'.

**The current behavior: is scrolling only the app bar height and stop scrolling in the webpage after!!**

```bash
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.14.5 18F132, locale en-EG)
    • Flutter version 1.22.3 at /Users/esham/Development/flutter
    • Framework revision 8874f21e79 (2 weeks ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3

 
[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/esham/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • ANDROID_HOME = /Users/esham/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/esham/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 39.0.3
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.51.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (1 available)
    • Redmi Note 8 Pro (mobile) • hev4t4bqo7hyvsnf • android-arm64 • Android 10 (API 29)
```

**Comment by PerLycke** [comment]
@blasten When you say fixed, more specifically what is fixed? In 1.22.4, with hybrid composition, you cannot use a CustomScrollView with a SliverAppBar and then a SliverFillRemaining with a webview. The webview just won't scroll

**Comment by heshaShawky** [comment]
This bug indeed still exists and not fixed yet!! I can't scroll on the webview itself and the sliverappbar even is glitchy, that in the latest version on flutter 1.22.4

> This has been fixed in Flutter 1.22/ WebView 1.0 - use hybrid composition in the Flutter WebView plugin.
> 
> https://github.com/flutter/plugins/tree/master/packages/webview_flutter#using-hybrid-composition

**Comment by edisonywh** [comment]
Hey team, any update on this or should we create another issue to track this? I tried it out with the new Hybrid Composition awhile back and that didn't seem to work. There is a new issue at https://github.com/flutter/flutter/issues/72888 but I believe that's a different issue.

The original problem as reported in this issue is to have a SliverAppBar + a WebView, and that scrolling on the WebView should collapse the SliverAppBar. The code example and a GIF that accompanies it describes it pretty well.

**Comment by blasten** [comment]
This issue is about the scroll performance while having (a) a platform view and (b) a `SilverAppBar`.   This issue was closed because hybrid composition did solve the root cause to the issue, which is the performance of resizing the webview. 

I hear that it would be useful to have an example that deals with other issues:
1) Nested content scrolling while a platform view is present.

I'd love to hear if any of you would be interested in sending a PR to the WebView plugin example app that accomplishes exactly this. Some ideas about how to approach it:

1. Use `WebViewController#evaluateJavascript` (For [example](https://github.com/flutter/plugins/blob/master/packages/webview_flutter/example/lib/main.dart#L220)) to get the `height` of the  HTML document. e.g. `document.scrollingElement.scrollHeight`.
2. Add a `SizedBox` as a child of `SliverFillRemaining`, and set the height to the value you received from the JS controller.

**Comment by emvaized** [comment]
I guess the best solution I came up with for now is:
1. Wrap webview with Scaffold which has empty preferred size widget as an appbar
2. Wrap webview with Listener widget, and update animation controller value on vertical movement
3. Change empty appbar height according to animation controller value
4. Display separate app bar widget in a stack on top of all these, and animate it with SlideTransition according to animation controller value

Here is quick snippet  for illustration: https://gist.github.com/emvaized/fa1ce586109196d7357652de2843b8ff

However, it has issues — for example app bar jumps back and forth when finger movement is very slow.
